producer/elasticsearch.go

```diff
@@ -15,297 +15,433 @@
-	originalMsg := msg.Clone()
+func (prod *ElasticSearch) submitMessages() core.AssemblyFunc {
+	return func(messages []*core.Message) {
+		client, err := prod.getClient()
+		if err != nil {
+			prod.Log.Error.Print("Sending messages failed: ", err)
+			return
+		}
+
+		bulkRequest := client.Bulk()
+		ctx := context.Background()
+
+		for _, msg := range messages {
+			indexMapItem := prod.indexMap[msg.GetStreamID()]
+
+			if indexMapItem.useDayIndex {
+				// create daily index ...
+				indexExists, _ := prod.isIndexExists(ctx, indexMapItem.name, client)
+				if !indexExists {
+					prod.createIndex(indexMapItem.name, indexMapItem.settings, indexExistsFalse)
+				}
+			}
+
+			bulkIndexRequest := elastic.NewBulkIndexRequest()
+			bulkIndexRequest.Index(indexMapItem.name).
+				Type(indexMapItem.typeName).
+				Doc(msg.String())
 
-	index, indexMapped := prod.index[msg.GetStreamID()]
-	if !indexMapped {
-		index, indexMapped = prod.index[core.WildcardStreamID]
-		if !indexMapped {
-			index = core.StreamRegistry.GetStreamName(msg.GetStreamID())
+			bulkRequest.Add(bulkIndexRequest)
+		}
+
+		// NumberOfActions contains the number of requests in a bulk
+		prod.Log.Debug.Printf("bulkRequest.NumberOfActions: %d", bulkRequest.NumberOfActions())
+
+		// Do sends the bulk requests to Elasticsearch
+		bulkResponse, err := bulkRequest.Do(ctx)
+		if err != nil {
+			prod.Log.Error.Print(err)
+		}
+
+		// Bulk request actions get cleared
+		numberOfActionsAfter := bulkRequest.NumberOfActions()
+		if numberOfActionsAfter != 0 {
+			prod.Log.Error.Printf("Could not send '%d' messages to Elasticsearch", numberOfActionsAfter)
+		}
+
+		if bulkResponse != nil {
+			// Indexed returns information abount indexed documents
+			indexed := bulkResponse.Indexed()
+			prod.Log.Debug.Printf("%d messages indexed successfully in Elasticsearch", len(indexed))
+
+			// Created returns information about created documents
+			created := bulkResponse.Created()
+			prod.Log.Debug.Printf("%d messages created successfully in Elasticsearch", len(created))
 		}
-		metricName := elasticMetricMessages + index
-		tgo.Metric.New(metricName)
-		tgo.Metric.NewRate(metricName, elasticMetricMessagesSec+index, time.Second, 10, 3, true)
-		prod.index[msg.GetStreamID()] = index
 	}
+}
 
-	indexType := prod.getIndexType(msg.GetStreamID())
+// -- elasticConnection --
+
+type elasticConnection struct {
+	servers           []string
+	user              string
+	password          string
+	setGzip           bool
+	client            *elastic.Client
+	isConnectedStatus bool
+	retrier           retrier
+}
 
-	sendIndex := index
-	if prod.dayBasedIndex {
-		sendIndex += msg.GetCreationTime().Format("_2006-01-02")
+func (conn *elasticConnection) isConnected() bool {
+	return conn.isConnectedStatus
+}
+
+func (conn *elasticConnection) connect() error {
+	conf := []elastic.ClientOptionFunc{elastic.SetURL(conn.servers...), elastic.SetSniff(false), elastic.SetGzip(conn.setGzip)}
+	if len(conn.user) > 0 {
```

Comment:
elastic search use a basic auth authentication - i don't see a possibility in elastic search to add credentials without a username: see https://www.elastic.co/guide/en/shield/current/native-realm.html#managing-native-users